NPE when variable is initialized using a method with the same name in features calling "thrown()"

# Issue description
A variable initialized using a method with the same name, as in `def foo = foo()` throws a `NullPointerException` when used in a feature method that calls `thrown()` in `then:` block.

# How to reproduce
```
class SomeSpec extends spock.lang.Specification {
    def "foo test"() { // fails
        when:
        def foo = foo() // fails here
        println(foo)
        throw new IllegalStateException()

        then:
        thrown IllegalStateException
    }

    def "bar test"() { // passes
        when:
        def bar = foo()
        println(bar)
        throw new IllegalStateException()

        then:
        thrown IllegalStateException
    }

    String foo() {
        return "foo"
    }
}
```

`foo test` fails when running this test class (only tried with Gradle):
```
Expected exception of type 'java.lang.IllegalStateException', but got 'java.lang.NullPointerException'
	at org.spockframework.lang.SpecInternals.checkExceptionThrown(SpecInternals.java:81)
	at org.spockframework.lang.SpecInternals.thrownImpl(SpecInternals.java:68)
	at SomeSpec.foo test(SomeSpec.groovy:9)
Caused by: java.lang.NullPointerException: Cannot invoke method call() on null object
	at SomeSpec.foo test(SomeSpec.groovy:4)
```

# Additional Environment information

## Java/JDK
`java -version`
```
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

## Build-tool dependencies used

### Gradle 6.8.1
Fails with: `testImplementation("org.spockframework:spock-core:2.0-M4-groovy-2.5")` and `testImplementation("org.spockframework:spock-core:2.0-M4-groovy-3.0")`

Works with: `testImplementation("org.spockframework:spock-core:1.3-groovy-2.5")`

--- a/spock-core/src/main/java/org/spockframework/compiler/SpecRewriter.java
+++ b/spock-core/src/main/java/org/spockframework/compiler/SpecRewriter.java
@@ -760,11 +760,26 @@ public class SpecRewriter extends AbstractSpecVisitor implements IRewriteResourc
           new BinaryExpression(
               copyLhsVariableExpressions(declExpr),
               Token.newSymbol(Types.ASSIGN, -1, -1),
-              declExpr.getRightExpression()));
+              transformRhsExpressionIfNecessary(declExpr)));
 
       declExpr.setRightExpression(createDefaultValueInitializer(declExpr));
       to.add(new ExpressionStatement(declExpr));
     }
+  }
+
+  private Expression transformRhsExpressionIfNecessary(DeclarationExpression declExpr) {
+    Expression rightExpression = declExpr.getRightExpression();
+    // The case of `def foo = foo()` will break if we split it up without changing the target,
+    // as groovy would now interpret it as a `foo.call()` invocation on the local variable.
+    if (rightExpression instanceof MethodCallExpression) {
+      MethodCallExpression methodCallExpression = (MethodCallExpression)rightExpression;
+      if (methodCallExpression.isImplicitThis() &&
+        declExpr.getVariableExpression().getName().equals(methodCallExpression.getMethod().getText())) {
+        // change to explicit `this` to turn the expression to `foo = this.foo()`
+        methodCallExpression.setImplicitThis(false);
+      }
+    }
+    return rightExpression;
   }
 
   private Expression createDefaultValueInitializer(DeclarationExpression expr) {

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/CleanupBlocks.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/CleanupBlocks.groovy
@@ -181,4 +181,20 @@ def feature() {
     assert d == 1d
     assert bool
   }
+
+  @Issue("https://github.com/spockframework/spock/issues/1266")
+  def "cleanup blocks don't destroy method reference when invocation is assigned to variable with the same name"() {
+    when:
+    def foobar = foobar()
+
+    then:
+    println(foobar)
+
+    cleanup:
+    foobar.size()
+  }
+
+  def foobar() {
+    return "foo"
+  }
 }

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/ast/CleanupBlocksAstSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/ast/CleanupBlocksAstSpec.groovy
@@ -1,0 +1,70 @@
+package org.spockframework.smoke.ast
+
+import org.spockframework.EmbeddedSpecification
+import spock.lang.Issue
+import spock.util.Show
+
+class CleanupBlocksAstSpec extends EmbeddedSpecification {
+
+  @Issue("https://github.com/spockframework/spock/issues/1266")
+  def "cleanup rewrite keeps correct method reference"() {
+    when:
+    def result = compiler.transpileSpecBody('''
+def "cleanup blocks don't destroy method reference when invocation is assigned to variable with the same name"() {
+  when:
+  def foobar = foobar()
+
+  then:
+  println(foobar)
+
+  cleanup:
+  foobar.size()
+}
+
+def foobar() {
+  return "foo"
+}''', EnumSet.of(Show.METHODS))
+    then:
+    result.source == '''\
+public java.lang.Object foobar() {
+    return 'foo'
+}
+
+public void $spock_feature_0_0() {
+    org.spockframework.runtime.ErrorCollector $spock_errorCollector = org.spockframework.runtime.ErrorRethrower.INSTANCE
+    org.spockframework.runtime.ValueRecorder $spock_valueRecorder = new org.spockframework.runtime.ValueRecorder()
+    java.lang.Object foobar
+    java.lang.Throwable $spock_feature_throwable
+    try {
+        foobar = this.foobar()
+        try {
+            org.spockframework.runtime.SpockRuntime.verifyMethodCondition($spock_errorCollector, $spock_valueRecorder.reset(), 'println(foobar)', 6, 3, null, this, $spock_valueRecorder.record($spock_valueRecorder.startRecordingValue(0), 'println'), new java.lang.Object[]{$spock_valueRecorder.record($spock_valueRecorder.startRecordingValue(1), foobar)}, $spock_valueRecorder.realizeNas(4, false), false, 3)
+        }
+        catch (java.lang.Throwable throwable) {
+            org.spockframework.runtime.SpockRuntime.conditionFailedWithException($spock_errorCollector, $spock_valueRecorder, 'println(foobar)', 6, 3, null, throwable)}
+        finally {
+        }
+    }
+    catch (java.lang.Throwable $spock_tmp_throwable) {
+        $spock_feature_throwable = $spock_tmp_throwable
+        throw $spock_tmp_throwable
+    }
+    finally {
+        try {
+            foobar.size()
+        }
+        catch (java.lang.Throwable $spock_tmp_throwable) {
+            if ( $spock_feature_throwable != null) {
+                $spock_feature_throwable.addSuppressed($spock_tmp_throwable)
+            } else {
+                throw $spock_tmp_throwable
+            }
+        }
+        finally {
+        }
+    }
+    this.getSpecificationContext().getMockController().leaveScope()
+}'''
+
+  }
+}

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/ast/condition/ExceptionConditionsAstSpec.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/ast/condition/ExceptionConditionsAstSpec.groovy
@@ -1,0 +1,46 @@
+package org.spockframework.smoke.ast.condition
+
+import org.spockframework.EmbeddedSpecification
+import spock.lang.Issue
+import spock.util.Show
+
+class ExceptionConditionsAstSpec extends EmbeddedSpecification {
+
+  @Issue("https://github.com/spockframework/spock/issues/1266")
+  def "thrown rewrite keeps correct method reference"() {
+    when:
+    def result = compiler.transpileSpecBody('''
+def "cleanup blocks don't destroy method reference when invocation is assigned to variable with the same name"() {
+  when:
+  def foobar = foobar()
+
+  then:
+  thrown(IllegalStateException)
+}
+
+def foobar() {
+  throw new IllegalStateException("foo")
+}''', EnumSet.of(Show.METHODS))
+    then:
+    result.source == '''\
+public java.lang.Object foobar() {
+    throw new java.lang.IllegalStateException('foo')
+}
+
+public void $spock_feature_0_0() {
+    java.lang.Object foobar
+    this.getSpecificationContext().setThrownException(null)
+    try {
+        foobar = this.foobar()
+    }
+    catch (java.lang.Throwable $spock_ex) {
+        this.getSpecificationContext().setThrownException($spock_ex)
+    }
+    finally {
+    }
+    this.thrownImpl(null, null, java.lang.IllegalStateException)
+    this.getSpecificationContext().getMockController().leaveScope()
+}'''
+
+  }
+}

--- a/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy
+++ b/spock-specs/src/test/groovy/org/spockframework/smoke/condition/ExceptionConditions.groovy
@@ -269,4 +269,17 @@ thrown() == e
     x == 3
     y == 4
   }
+
+  @Issue("https://github.com/spockframework/spock/issues/1266")
+  def "thrown conditions don't destroy method reference when invocation is assigned to variable with the same name"() {
+    when:
+    def foobar = foobar()
+
+    then:
+    thrown(IllegalStateException)
+  }
+
+  def foobar() {
+    throw new IllegalStateException("foo")
+  }
 }

